src/inference/src/dev/itensor.hpp

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include <memory>
+
+#include "openvino/core/coordinate.hpp"
+#include "openvino/core/core_visibility.hpp"
+#include "openvino/core/shape.hpp"
+#include "openvino/core/type/element_type.hpp"
+#include "openvino/runtime/allocator.hpp"
+
+namespace InferenceEngine {
+
+class Blob;
+
+}
+
+namespace ov {
+
+class ITensor : public std::enable_shared_from_this<ITensor> {
+public:
+    /**
+     * @brief Set new shape for tensor, deallocate/allocate if new total size is bigger than previous one.
+     * @note Memory allocation may happen
+     * @param shape A new shape
+     */
+    virtual void set_shape(const ov::Shape& shape);
+
+    /**
+     * @return A tensor element type
+     */
+    virtual const element::Type& get_element_type() const;
+
+    /**
+     * @return A tensor shape
+     */
+    virtual const Shape& get_shape() const;
+
+    /**
+     * @brief Returns the total number of elements (a product of all the dims or 1 for scalar)
+     * @return The total number of elements
+     */
+    virtual size_t get_size() const;
+
+    /**
+     * @brief Returns the size of the current Tensor in bytes.
+     * @return Tensor's size in bytes
+     */
+    virtual size_t get_byte_size() const;
+
+    /**
+     * @return Tensor's offseta in elements
+     */
+    virtual Coordinate get_offsets() const;
+
+    /**
+     * @return Tensor's strides in bytes
+     */
+    virtual Strides get_strides() const;
+
+    /**
+     * @brief Provides an access to the underlaying host memory
+     * @param type Optional type parameter.
+     * @note If type parameter is specified, the method throws an exception
+     * if specified type's fundamental type does not match with tensor element type's fundamental type
+     * @return A host pointer to tensor memory
+     */
+    virtual void* data(const element::Type& type = {}) const;
+
+    /**
+     * @brief Provides an access to the underlaying host memory casted to type `T`
+     * @return A host pointer to tensor memory casted to specified type `T`.
+     * @note Throws exception if specified type does not match with tensor element type
+     */
+    template <typename T, typename datatype = typename std::decay<T>::type>
+    T* data() const {
+        return static_cast<T*>(data(element::from<datatype>()));
+    }
+
+    /**
+     * @brief Return additional information associated with tensor
+     * @return Map of property names to properties
+     */
+    virtual AnyMap get_properties() const;
+
+protected:
+    ~ITensor() = default;
+};
+
+/**
+ * @brief Constructs Tensor using element type and shape. Allocate internal host storage using default allocator
+ * @param type Tensor element type
+ * @param shape Tensor shape
+ * @param allocator allocates memory for internal tensor storage
+ */
+std::shared_ptr<ITensor> make_tensor(const element::Type type, const Shape& shape, const Allocator& allocator = {});
```

Comment:
Maybe these tensor makers should be template and support arguments forwarding in case when tensor created from temporaries then parameters can be forwarded (moved) into created object?